### PR TITLE
fix(cloud_firestore): startAfterDocument with where a DocumentReference throws

### LINKED ISSUE #10338: 🐛 [cloud_firestore] startAfterDocument with where a DocumentReference throws
## Bug report

**Describe the bug**
When having a query like this (see example project)
```
final documentSnapshot = await ff.doc('/movies/1').get();
final query = ff
    .collection('movies')
    .where('ref', isNull: false) // this is document reference
    .orderBy('ref')
    .limit(10)
    .startAfterDocument(documentSnapshot);
res = (await query.snapshots().first).toString();
```

Executing this query will throw.

```
flutter: Error: Invalid argument: Instance of ‘_JsonDocumentReference’
Stack: #0      StandardMessageCodec.writeValue (package:flutter/src/services/message_codecs.dart:466:7)
#1      FirestoreMessageCodec.writeValue (package:cloud_firestore_platform_interface/src/method_channel/utils/firestore_message_codec.dart:119:13)
#2      StandardMessageCodec.writeValue (package:flutter/src/services/message_codecs.dart:456:9)
#3      FirestoreMessageCodec.writeValue (package:cloud_firestore_platform_interface/src/method_channel/utils/firestore_message_codec.dart:119:13)
#4      StandardMessageCodec.writeValue (package:flutter/src/services/message_codecs.dart:456:9)
#5      FirestoreMessageCodec.writeValue (package:cloud_firestore_platform_interface/src/method_channel/utils/firestore_message_codec.dart:119:13)
#6      StandardMessageCodec.writeValue.<anonymous closure> (package:flutter/src/services/message_codecs.dart:463:9)
#7      _LinkedHashMapMixin.forEach (dart:collection-patch/compact_hash.dart:617:1<…>
flutter: ----------------FIREBASE CRASHLYTICS----------------
flutter: Invalid argument: Instance of ‘_JsonDocumentReference’
flutter:
#0      StandardMessageCodec.writeValue (package:flutter/src/services/message_codecs.dart:466:7)
#1      FirestoreMessageCodec.writeValue (package:cloud_firestore_platform_interface/src/method_channel/utils/firestore_message_codec.dart:119:13)
#2      StandardMessageCodec.writeValue (package:flutter/src/services/message_codecs.dart:456:9)
#3      FirestoreMessageCodec.writeValue (package:cloud_firestore_platform_interface/src/method_channel/utils/firestore_message_codec.dart:119:13)
#4      StandardMessageCodec.writeValue (package:flutter/src/services/message_codecs.dart:456:9)
#5      FirestoreMessageCodec.writeValue (package:cloud_firestore_platform_interface/src/method_channel/utils/firestore_message_codec.dart:119:13)
#6      StandardMessageCodec.writeValue.<anonymous closure> (package:flutter/src/services/message_codecs.dart:463:9)
#7      _LinkedHashMapMixin.forEach (dart:collection-patch/compact_hash.dart:617:13)
#8      MapView.forEach (dart:collection/maps.dart:342:10)
#9    <…>
flutter: ----------------------------------------------------
flutter: ----------------FIREBASE CRASHLYTICS----------------
flutter: Invalid argument: Instance of ‘_JsonDocumentReference’
flutter:
#0      StandardMessageCodec.writeValue (package:flutter/src/services/message_codecs.dart:466:7)
#1      FirestoreMessageCodec.writeValue (package:cloud_firestore_platform_interface/src/method_channel/utils/firestore_message_codec.dart:119:13)
#2      StandardMessageCodec.writeValue (package:flutter/src/services/message_codecs.dart:456:9)
#3      FirestoreMessageCodec.writeValue (package:cloud_firestore_platform_interface/src/method_channel/utils/firestore_message_codec.dart:119:13)
#4      StandardMessageCodec.writeValue (package:flutter/src/services/message_codecs.dart:456:9)
#5      FirestoreMessageCodec.writeValue (package:cloud_firestore_platform_interface/src/method_channel/utils/firestore_message_codec.dart:119:13)
#6      StandardMessageCodec.writeValue.<anonymous closure> (package:flutter/src/services/message_codecs.dart:463:9)
#7      _LinkedHashMapMixin.forEach (dart:collection-patch/compact_hash.dart:617:13)
#8      MapView.forEach (dart:collection/maps.dart:342:10)
#9    <…>
flutter: ----------------------------------------------------
```

### Steps to reproduce

1. See sample project.
2. Possible fix branch is commented out in `pubspec.yaml`

### Expected behavior

The query to return successfully. 

### Sample project

https://github.com/agent3bood/FlutterFireBugReport

---

### Flutter doctor

Run `flutter doctor` and paste the output below:

<details><summary>Click to expand</summary>

```
❯ flutter doctor
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 3.3.9, on macOS 12.6 21G115 darwin-arm, locale en-TR)
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
[✓] Xcode - develop for iOS and macOS (Xcode 14.0)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2021.1)
[✓] Android Studio (version 2021.1)
[✓] IntelliJ IDEA Ultimate Edition (version 2021.3.3)
[✓] IntelliJ IDEA Ultimate Edition (version 2021.3.3)
[✓] VS Code (version 1.69.2)
[✓] Connected device (1 available)
[✓] HTTP Host Availability

• No issues found!
```

</details>

---

### Flutter dependencies

Run `flutter pub deps -- --style=compact` and paste the output below:

<details><summary>Click To Expand</summary>

```
❯ flutter pub deps -- --style=compact
Dart SDK 2.18.5
Flutter SDK 3.3.9
flutter_fire_bug_1 1.0.0+1

dependencies:
- cloud_firestore 4.3.1 [cloud_firestore_platform_interface cloud_firestore_web collection firebase_core firebase_core_platform_interface flutter meta]
- cupertino_icons 1.0.5
- firebase_core 2.4.1 [firebase_core_platform_interface firebase_core_web flutter meta]
- flutter 0.0.0 [characters collection material_color_utilities meta vector_math sky_engine]

dev dependencies:
- flutter_lints 2.0.1 [lints]
- flutter_test 0.0.0 [flutter test_api path fake_async clock stack_trace vector_math async boolean_selector characters collection matcher material_color_utilities meta source_span stream_channel string_scanner term_glyph]

transitive dependencies:
- _flutterfire_internals 1.0.12 [collection firebase_core firebase_core_platform_interface flutter meta]
- async 2.9.0 [collection meta]
- boolean_selector 2.1.0 [source_span string_scanner]
- characters 1.2.1
- clock 1.1.1
- cloud_firestore_platform_interface 5.10.1 [_flutterfire_internals collection firebase_core flutter meta plugin_platform_interface]
- cloud_firestore_web 3.2.1 [_flutterfire_internals cloud_firestore_platform_interface collection firebase_core firebase_core_web flutter flutter_web_plugins js]
- collection 1.16.0
- fake_async 1.3.1 [clock collection]
- firebase_core_platform_interface 4.5.2 [collection flutter flutter_test meta plugin_platform_interface]
- firebase_core_web 2.1.0 [firebase_core_platform_interface flutter flutter_web_plugins js meta]
- flutter_web_plugins 0.0.0 [flutter js characters collection material_color_utilities meta vector_math]
- js 0.6.4
- lints 2.0.1
- matcher 0.12.12 [stack_trace]
- material_color_utilities 0.1.5
- meta 1.8.0
- path 1.8.2
- plugin_platform_interface 2.1.3 [meta]
- sky_engine 0.0.99
- source_span 1.9.0 [collection path term_glyph]
- stack_trace 1.10.0 [path]
- stream_channel 2.1.0 [async]
- string_scanner 1.1.1 [source_span]
- term_glyph 1.2.1
- test_api 0.4.12 [async boolean_selector collection meta source_span stack_trace stream_channel string_scanner term_glyph matcher]
- vector_math 2.1.2

```

</details>

---

### PR DESCRIPTION
**Pull request by agent3bood**
## Description

*Replace this paragraph with a description of what this PR is doing. If you're modifying existing behavior, describe the existing behavior, how this PR is changing it, and what motivated the change.*

## Related Issues

closes #10338

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process. Updating the `pubspec.yaml` and changelogs is not required.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] The analyzer (`melos run analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change.
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/firebase/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

The command `melos run analyze` fails in packages not related to my PR, see bellow.
```
$ melos exec
  └> dart analyze . --fatal-infos
     └> FAILED (in 4 packages)
        └> cloud_firestore_odm (with exit code 1)
        └> cloud_firestore_odm_example (with exit code 1)
        └> cloud_firestore_odm_generator (with exit code 1)
        └> cloud_firestore_odm_generator_integration_test (with exit code 1)

```

### CODE REVIEW

File: packages/cloud_firestore/cloud_firestore/test/query_test.dart [hunk 1]
```diff
@@ -309,6 +336,20 @@ void main() {
...
+            await firestore.collection('/movies').doc('exists').get();
+        Query q = query!.orderBy('reference').startAfterDocument(doc);
+        expect(
+          q.parameters['startAfter'].first,
+          allOf([
+            isA<MethodChannelDocumentReference>(),
+            isA<DocumentReferencePlatform>(),
+          ]),
+        );
+      });
```
Comment (Lyokone) [inline packages/cloud_firestore/cloud_firestore/test/query_test.dart:352]: Could you move this test to the integration tests: `packages/cloud_firestore/cloud_firestore/example/integration_test/query_e2e.dart`? So you don't need to mock the channel and we can check it works on a real device.

**Review by Lyokone — CHANGES_REQUESTED** [review]
Looks good but tests should be moved to integration testing.

**Review by Lyokone — APPROVED** [review]

**Comment by Lyokone** [comment]
Rebased the branch on master to get CI working with Flutter 3.7

**Review by russellwheatley — APPROVED** [review]

**Comment by Lyokone** [comment]
Thanks for your help :)

**Pull request merged**